### PR TITLE
[User History] Track User delete

### PR DESCRIPTION
**Pull request by mkangia**
## Summary

:fish: / :fish: 
Builds on the [base setup](https://github.com/dimagi/commcare-hq/pull/29887) to track user deletion.
This PR also adds a way to skip passing domain for deletions done by system admins where there is no domain.

## Product Description

No UI changes.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### QA Plan

same as base PR

### Safety story

This was first run of master to run tests and only after a [green build](https://travis-ci.com/github/dimagi/commcare-hq/builds/228648661), base was changed to mk/user-history/base

### Rollback instructions

same as base PR

### CODE REVIEW

File: corehq/apps/auditcare/tests/test_auth.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def create_web_user(self):
         # A Couch user is required for logout
         user = WebUser.create('', USERNAME, 'mockmock', None, None)
         user.save()
-        self.addCleanup(lambda: user.delete(None))
+        self.addCleanup(user.delete, '', deleted_by=None)
```
Comment (snopoke) [inline corehq/apps/auditcare/tests/test_auth.py:31]: why not just pass in `None` instead of a blank string?

File: corehq/apps/cleanup/tasks.py [hunk 2]
```diff
@@ -253,7 +254,7 @@ def delete_web_user():
         ]:
             web_user = WebUser.get_by_username(username)
             if web_user:
-                web_user.delete(deleted_by=None, deleted_via=__name__)
+                web_user.delete(None, deleted_by=SYSTEM_USER_ID, deleted_via=__name__ + '.delete_web_user')
```
Comment (snopoke) [inline corehq/apps/cleanup/tasks.py:257]: won't this fail the `if not domain` validation?

File: corehq/apps/users/models.py [hunk 3]
```diff
@@ -1169,16 +1170,16 @@ def user_session_data(self):
...
         self.clear_quickcache_for_user()
         try:
             user = self.get_django_user()
             user.delete()
-            if deleted_by:
-                log_model_change(deleted_by, user, message=f"deleted_via: {deleted_via}",
-                                 action=ModelAction.DELETE)
         except User.DoesNotExist:
             pass
+        if deleted_by:
```
Comment (snopoke) [inline corehq/apps/users/models.py:1180]: similar comment to on the other PR, is this check just for tests? Should it validate that `deleted_by` is None only in tests.

Another thought, if this is being skipped in tests what's the point of passing in the domain in all the test usages? Why not just pass in `None`

File: corehq/apps/cleanup/tasks.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/cleanup/tasks.py:257]: I see you changed the method to allow `None` for domain if the user is `SYSTEM_USER_ID`

**Review by snopoke — APPROVED** [review]

File: corehq/apps/api/resources/v0_5.py [hunk 4]
```diff
@@ -373,16 +373,15 @@ def obj_create(self, bundle, **kwargs):
...
+                                  deleted_via=USER_CHANGE_VIA_API)
             else:
                 try:
                     django_user = bundle.obj.get_django_user()
                 except User.DoesNotExist:
                     pass
                 else:
                     django_user.delete()
-                    log_model_change(bundle.request.user, django_user, message=f"deleted_via: {USER_CHANGE_VIA_API}",
-                                     action=ModelAction.DELETE)
```
Comment (mkangia) [inline corehq/apps/api/resources/v0_5.py:385]: Since we are logging in delete above, this should not be needed.

File: corehq/apps/auditcare/tests/test_auth.py (hunk 1 above)
Comment (mkangia) [reply corehq/apps/auditcare/tests/test_auth.py:31]: We could. Just that the domain in this test is a blank string and i was simply passing domain in all tests. Not sure why we even allow that.

File: corehq/apps/users/models.py (hunk 3 above)
Comment (mkangia) [reply corehq/apps/users/models.py:1180]: > similar comment to on the other PR

adding for reference, https://github.com/dimagi/commcare-hq/pull/29902#discussion_r650870658

> is this check just for tests? Should it validate that deleted_by is None only in tests.

yes. adding a check to ensure deleted_by is only for tests.


> Another thought, if this is being skipped in tests what's the point of passing in the domain in all the test usages? Why not just pass in None

You could pass None yes.
I think I went with domain because the original intention was to "avoid logging in tests to avoid DB overhead" and not to "allow tests to not pass domain".

Open to suggestions on this. 

Honestly, 
When i was fixing tests and passing domain, it seemed okay to do so but when it was too many files and i was regretting it, it was too late, so I just continued on it. Was thinking that I could have just updated the method signature to have `domain=None` and not worry about it in tests.

Once done, was thinking about positives,
1. that there was not one more `None` next to `deleted_by=None` in tests, which was nicer to read.
2. this built a sense that domain is to be passed on user actions now and get that as a habit, which is going to be new for everyone.

Comment (snopoke) [reply corehq/apps/users/models.py:1182]: I think it's fine as is and no need to go back and change all places to pass None

**Review by snopoke — APPROVED** [review]

File: corehq/apps/users/models.py (hunk 3 above)
Comment (mkangia) [reply corehq/apps/users/models.py:1182]: What would you suggest the ideal approach would have been here?

File: corehq/apps/cleanup/tasks.py (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/cleanup/tasks.py:257]: Why not support `deleted_by` being None for system changes instead of checking for `SYSTEM_USER_ID`? It's weird that `deleted_by` can be either a user or a string, and it's unfortunate to add that bit of complexity that's only used by this one caller.

Tangentially, this is a weird task. The original PR (https://github.com/dimagi/commcare-hq/pull/23551) says this task was requested by QA. @Ben-Talbot do you know what this task is and if it's still needed? It checks every night if there's a web user named either `create_growth<at>outlook.com` or `growth_analytics<at>outlook.com`, and if it finds such a user, deletes it.

**Review by orangejenny — APPROVED** [review]

File: corehq/apps/cleanup/tasks.py (hunk 2 above)
Comment (mkangia) [reply corehq/apps/cleanup/tasks.py:257]: > Why not support deleted_by being None for system changes instead of checking for SYSTEM_USER_ID?

You mean assume that if there is no deleted by, it was system. I am not sure I like that, being explicit is good.
But also, currently we recognize it's a system change eventually from the `deleted_by` passed, so if that is not passed, we wont be able to identify such changes.

>  can be either a user or a string

I think HQ treats `SYSTEM_USER_ID` as an ID [along with some other ones](https://github.com/dimagi/commcare-hq/blob/f8f382e06a90820dc44d21998da9a95927f7af9d/corehq/pillows/utils.py#L59). Its the same thing we have with forms as well.

>  bit of complexity that's only used by this one caller.

Its true that this is one instance and would be great if we can stop doing it. Though I think we would have more such system updates to user in future PR so would be good to agree on a way to log updates to users by SYSTEM_USER_ID.

Comment (orangejenny) [reply corehq/apps/cleanup/tasks.py:257]: I do agree with you about it being good to be explicit. I'm surprised there aren't other places where we have the system changing users. You've gone through all of the code, right? Maybe we just don't often have the system *delete* users.

> I think HQ treats SYSTEM_USER_ID as an ID

Even though it's a known ID, it's still a string, so `changed_by_user` is either a string or a couch document, which are two very different types to be represented by the same variable.

File: corehq/apps/users/models.py (hunk 3 above)
Comment (snopoke) [reply corehq/apps/users/models.py:1182]: probably just passing None in tests

**Comment by mkangia** [comment]
Going to merge this in base PR branch, please let me know if there is any more feedback on comments on this, I will follow up on base PR.

**Comment by crimsonsun25** [comment]
> Tangentially, this is a weird task. The original PR (#23551) says this task was requested by QA. @Ben-Talbot do you know what this task is and if it's still needed? It checks every night if there's a web user named either create_growth<at>outlook.com or growth_analytics<at>outlook.com, and if it finds such a user, deletes it. 

Responding on the QA query here - The user deletion activity added in the linked PR was done for [this task](https://dimagi-dev.atlassian.net/browse/QA-47) of Growth Team. That task is now dead and if required, feel free to purge that PR as well
@orangejenny @mkangia

**Comment by mkangia** [comment]
thank you @crimsonsun25. Will take the necessary action to remove the task.
